### DISCUSSION TITLE
Unable to install different flavours of SunbirdEd-mobile app in same phone for android

### DISCUSSION THREAD

**Discussion by kiranharidas187** [Sunbird-Ed/Community#12] [Installation] [answered]
@sharathkashyap @luckynvdu 

We have taken the latest code from the release-4.6.0 branch and created a new flavor of the android app(called APEX) . On testing, we have identified that the Apex app can't be installed on the phone having the Diksha app and vice versa. This same issue happens between staging, pre-prod, and prod build of Diksha also. On further investigation, we found that the issue is with cordova-plugin-document-viewer.

**Error while trying to install**

`adb: failed to install Downloads/diksha-4.8.990preproduction-debug.apk: Failure [INSTALL_FAILED_CONFLICTING_PROVIDER: Scanning Failed.: Can't install because provider name org.sunbird.app.DocumentViewerPlugin.fileprovider (in package preprod.diksha.app) is already used by org.elevate.apex.staging]`

**Cause for issue**
In AndroidManifets.xml
` <provider
            android:name="org.apache.cordova.camera.FileProvider"
            android:authorities="org.elevate.apex.cordova.plugin.camera.provider"
            android:exported="false"
            android:grantUriPermissions="true" >
            <meta-data
                android:name="android.support.FILE_PROVIDER_PATHS"
                android:resource="@xml/camera_provider_paths" />
        </provider>
        <provider
            android:name="de.sitewaerts.cordova.documentviewer.FileProvider"
            android:authorities="org.sunbird.app.DocumentViewerPlugin.fileprovider"
            android:exported="false"
            android:grantUriPermissions="true" >
            <meta-data
                android:name="android.support.FILE_PROVIDER_PATHS"
                android:resource="@xml/documentviewer_file_paths" />
        </provider>`

Only for the Document viewer plugin, the appId is not taken from env variables instead it's taken from config.xml.

**Solutions tried**
- Changed the packaged in config.xml and tried to create a build. But in that case, got the following error 
   

```
Task :app:compileDevDebugJavaWithJavac FAILED
/Volumes/disk0s2/workspace/APEX/SunbirdEd-mobile-app/platforms/android/app/src/main/java/org/sunbird/SplashScreen.java:36: error: cannot find symbol
import org.sunbird.app.R;
                      ^
  symbol:   class R
  location: package org.sunbird.app
/Volumes/disk0s2/workspace/APEX/SunbirdEd-mobile-app/platforms/android/app/src/main/java/com/gae/scaffolder/plugin/MyFirebaseMessagingService.java:25: error: cannot find symbol
import org.sunbird.app.R;
                      ^
  symbol:   class R
  location: package org.sunbird.app
/Volumes/disk0s2/workspace/APEX/SunbirdEd-mobile-app/platforms/android/app/src/main/java/org/sunbird/SplashScreen.java:497: error: package R does not exist
        splashImageBottomView.setImageResource(R.drawable.ic_combined_shape);
                                                ^
/Volumes/disk0s2/workspace/APEX/SunbirdEd-mobile-app/platforms/android/app/src/main/java/com/gae/scaffolder/plugin/MyFirebaseMessagingService.java:246: error: package R does not exist
                    .setSmallIcon(R.mipmap.n_icon)
                                   ^
/Volumes/disk0s2/workspace/APEX/SunbirdEd-mobile-app/platforms/android/app/src/main/java/com/gae/scaffolder/plugin/MyFirebaseMessagingService.java:276: error: package R does not exist
                    .setSmallIcon(R.mipmap.n_icon)
                                   ^
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
5 errors

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileDevDebugJavaWithJavac'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org/

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.5/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 10s
33 actionable tasks: 32 executed, 1 up-to-date
Command failed with exit code 1: /Volumes/disk0s2/workspace/APEX/SunbirdEd-mobile-app/platforms/android/gradlew cdvBuildDebug -b /Volumes/disk0s2/workspace/APEX/SunbirdEd-mobile-app/platforms/android/build.gradle
[ERROR] An error occurred while running subprocess Cordova.
        
        cordova build android exited with exit code 1.
        
        Re-running this command with the --verbose flag may provide more information.

```
- Went through the document viewer plugin to understand if there is any way to update the packageId, but could not find any solution.

We are holding back the release of our app as the users having the Diksha app on their phones will not be able to install the Apex app. 

cc- @vijiurs

**Comment by sharathkashyap** [comment, marked as answer]
@kiranharidas187 This has been limitation introduced since document-viewer plugin . It tries to register based on packageID during installation. There can be two solutions to this:
a) Either take a fork of plugin where it tries to register itself under a dedicated package as prefix.
b) We may have to remove the dependency from code base for this plugin. (i.e. trying to find alternate flows where this plugin is used).

@luckynvdu This is a valid tech debt for the Mobile App

**Reply by pramodkvarma** [reply]
Is the package name provided during build? If so, could we prefix/suffix version number to package name in build file and the  build the mobile app? That should allow two different versions of mobile app, wouldn't it? If this works, should be an easy build script fix, right? @sharathkashyap

**Reply by swayangjit** [reply]
@pramodkvarma The document viewer plugin creates the **authorities** name based on the package name of the app. Ideally, it should be based on the **applicationId** which is a common convention for plugins that uses ContentProvider. I have forked the plugin and tweaked it to use the **applicationId** using which we can install multiple versions of mobile app. As of now it's available in my fork, we will move it to the Sunbirded repo if everything works fine.
@kiranharidas187 You can use [this plugin](https://github.com/swayangjit/cordova-plugin-document-viewer) and let us know if that is working or not.

**Reply by sharathkashyap** [reply]
@pramodkvarma This is provided in plugin.xml by the original publisher of the plugin.
https://github.com/sitewaerts/cordova-plugin-document-viewer/blob/c1eb0ac8fd2d75a539349b79766d213f47674a6d/plugin.xml#L52

`android:authorities="$PACKAGE_NAME.DocumentViewerPlugin.fileprovider"` 

Most of the plugin providers would keep it on packageID instead thus avoiding the conflict.
@swayangjit will be creating a separate fork of this plugin and we will do the needful change and get it tested.

**Reply by sharathkashyap** [reply]
@kiranharidas187 We have created a dedicated Jenkins Job for creating Multiple WhiteLabel Applications. Please feel free to use it rather than creating application from scratch with all configs. Will share more details shortly on this.

cc: @luckynvdu @swayangjit

**Reply by pramodkvarma** [reply]
Excellent! Thanks @swayangjit and @sharathkashyap .
